swift/common/middleware/ephemeral_object.py

```diff
@@ -0,0 +1,385 @@
+            for name in entries:
+                yield name
+            if len(entries) < limit:
+                break
+
+            # the last entry was a prefix, use it as marker to use next entry
+            # as it may have lot of items using this prefix
+            # (using exclusive start)
+            subdir_prefix = self._extract_subdir_prefix(name)
+            if subdir_prefix is None:
+                min = '(' + name
+            else:
+                min = '(' + subdir_prefix + FAKE_MARKER
+        while True:
+            yield None
+
+    # END: _build_object_listing
+
+    def _handle_mpu_object(self, req, start_response, account, container,
+                           path):
+        """Handler uploadId requests on memory backend.
+           Return tuple (response, fallback_needed)
+        """
+        lkey, hkey = self.keys(account, container)
+        if req.method in ('GET', 'HEAD'):
+            is_available = self.conn.hexists(hkey, path)
+            if is_available:
+                res = self.conn.hget(hkey, path)
+                oheaders = json.loads(res) if res else {}
+                start_response("200 OK", oheaders.items())
+            else:
+                start_response("404 Not Found", [])
+            return [b'']
+
+        if req.method == 'PUT':
+            hdrs = {k: v for k, v in req.headers.items()}
+            hdrs['last_modified'] = time.strftime(
+                "%Y-%m-%dT%H:%M:%S.000000", time.gmtime())
+            # Add versioning set by OIO
+            hdrs['x-amz-version-id'] = int(time.time() * 1e6)
+            self._create_entry((lkey, hkey), path, hdrs)
+            oheaders = {'Content-Length': 0,
+                        'Etag': MD5_OF_EMPTY_STRING}
+            start_response("201 Created", oheaders.items())
+            # The new uploadId are only managed on memory backend
+            return [b'']
+
+        if req.method == 'DELETE':
+            is_available = self.conn.hexists(hkey, path)
+            # not sure if it is needed as abort-multipart-upload
+            # begins with a test on uploadId validity
+            if not is_available:
+                start_response("404 Not Found", [])
+                return [b'']
+            self._remove_entry((lkey, hkey), path)
+
+            oheaders = {'Content-Length': 0,
+                        'Etag': MD5_OF_EMPTY_STRING}
+            start_response("204 No Content", oheaders.items())
+            return [b'']
+
+    def __call__(self, env, start_response):
+        req = Request(env)
+        account, container, obj = self._extract_path(req.path_info)
+
+        if container is None or '+segments' not in container:
+            return self.app(env, start_response)
+
+        # if obj and prefix are None with container+segments, we want the
+        # normal listing because it is the list-multipart-uploads operation
+        # reserved for MPU uploadId
+        if req.environ.get('oio.list_mpu'):
```

Comment:
We should use only one environment variable : `oio.ephemeral_object`
We can know if it is a listing thanks to the method used and to the presence of an object in the path.